psi4/src/psi4/dlpno/mp2.cc

```diff
@@ -0,0 +1,1413 @@
+
+        // value of mo at each point squared
+        C_lmo_slice = linalg::doublet(point_values_trim, C_lmo_slice, false, false);  // points x naocc
+        C_pao_slice = linalg::doublet(point_values_trim, C_pao_slice, false, false);  // points x npao
+
+        for (size_t p = 0; p < npoints_block; p++) {
+            for (size_t i = 0; i < naocc; ++i) {
+                C_lmo_slice->set(p, i, pow(C_lmo_slice->get(p, i), 2));
+            }
+            for (size_t u = 0; u < npao; ++u) {
+                C_pao_slice->set(p, u, pow(C_pao_slice->get(p, u), 2));
+            }
+        }
+
+        SharedMatrix C_lmo_slice_w = std::make_shared<Matrix>(C_lmo_slice);  // points x naocc
+        for (size_t p = 0; p < npoints_block; p++) {
+            C_lmo_slice_w->scale_row(0, p, block->w()[p]);
+        }
+
+        DOI_ij_temps[thread]->add(linalg::doublet(C_lmo_slice_w, C_lmo_slice, true, false));  // naocc x naocc
+        DOI_iu_temps[thread]->add(linalg::doublet(C_lmo_slice_w, C_pao_slice, true, false));  // naocc x npao
+    }
+    timer_off("Integration");
+
+    DOI_ij = std::make_shared<Matrix>("(i,j) Differential Overlap Integrals", naocc, naocc);
+    DOI_iu = std::make_shared<Matrix>("(i,u) Differential Overlap Integrals", naocc, nbf);
+
+    for (size_t thread = 0; thread < nthread; thread++) {
+        DOI_ij->add(DOI_ij_temps[thread]);
+        DOI_iu->add(DOI_iu_temps[thread]);
+    }
+
+    DOI_ij->sqrt_this();
+    DOI_iu->sqrt_this();
+}
+
+void DLPNOMP2::dipole_ints() {
+    int natom = molecule_->natom();
+    int naocc = C_lmo->colspi(0);
+    int nbf = C_lmo->rowspi(0);
+
+    std::shared_ptr<MintsHelper> mints = std::make_shared<MintsHelper>(basisset_, options_);
+    std::vector<SharedMatrix> ao_dipole = mints->ao_dipole();
+
+    SharedMatrix lmo_lmo_dipx = linalg::triplet(C_lmo, ao_dipole[0], C_lmo, true, false, false);
+    SharedMatrix lmo_lmo_dipy = linalg::triplet(C_lmo, ao_dipole[1], C_lmo, true, false, false);
+    SharedMatrix lmo_lmo_dipz = linalg::triplet(C_lmo, ao_dipole[2], C_lmo, true, false, false);
+
+    SharedMatrix lmo_pao_dipx = linalg::triplet(C_lmo, ao_dipole[0], C_pao, true, false, false);
+    SharedMatrix lmo_pao_dipy = linalg::triplet(C_lmo, ao_dipole[1], C_pao, true, false, false);
+    SharedMatrix lmo_pao_dipz = linalg::triplet(C_lmo, ao_dipole[2], C_pao, true, false, false);
+
+    // < i | dipole | i >
+    std::vector<Vector3> R_i;
+
+    // < i | dipole | u >
+    std::vector<std::vector<Vector3>> lmo_pao_dr(naocc);
+
+    // e_u
+    std::vector<SharedVector> lmo_pao_e(naocc);
+
+    for (size_t i = 0; i < naocc; ++i) {
+        R_i.push_back(Vector3(lmo_lmo_dipx->get(i, i), lmo_lmo_dipy->get(i, i), lmo_lmo_dipz->get(i, i)));
+    }
+
+    for (size_t i = 0; i < naocc; ++i) {
+        std::vector<int> pao_inds;
+        for (size_t u = 0; u < nbf; u++) {
+            if (fabs(DOI_iu->get(i, u)) > options_.get_double("T_CUT_DO_PRE")) {
```

Comment:
I'd cache the value of `T_CUT_DO_PRE` outside the naocc for loop otherwise you're doing a lookup naocc * nbf times.